cpp/ql/src/Security/CWE/CWE-120/UnboundedWrite.ql

```diff
@@ -73,11 +73,42 @@ predicate isSink(DataFlow::Node sink, BufferWrite bw, boolean qualifier) {
   unboundedWriteSource(sink.asDefiningArgument(), bw, qualifier)
 }
 
+/**
+ * A configuration that specifies flow from a `FlowSource` to a relational
+ * comparison.
+ *
+ * This configuration is used to speed up the barrier computations in Config.
+ */
+module BarrierConfig implements DataFlow::ConfigSig {
+  predicate isSource(DataFlow::Node source) { isSource(source, _) }
+
+  predicate isSink(DataFlow::Node sink) {
+    comparesEq(sink.asOperand(), _, _, true, _) or
+    comparesLt(sink.asOperand(), _, _, true, _)
+  }
+}
+
+module BarrierFlow = TaintTracking::Global<BarrierConfig>;
+
+import semmle.code.cpp.ir.dataflow.internal.DataFlowImplCommon as DataFlowImplCommon
+
+/**
+ * Holds if `left` is a left operand of some relational comparison that may
+ * depend on user input.
+ */
+predicate interestingLessThanOrEqual(Operand left) {
+  exists(DataFlowImplCommon::NodeEx node |
+    node.asNode().asOperand() = left and
+    BarrierFlow::Stages::Stage1::sinkNode(node, _)
```

Comment:
So my understanding/guessing is that stage one does some cheap flow starting at sources to find sink that could be interesting, and we are using that capability to prune down barriers as well.

Is this something that the data flow library could actually do itself? Prune down barriers in this way. Or could it be bad to do more generally?

Reply:
The problem is that the `lessThanOrEqual` predicate itself blows up due to the fanout from `g` in `g.comparesLt(left, _, _, true, branch)` (or `g.comparesEq(left, _, _, true, branch)`) on some projects containing a function with many guards that guard equivalent expressions (so that there are many guards for the same `left`).

As an alternative we could've inlined `lessThanOrEqual` and `isBarrier`, and then we'd most likely get to a point where the resulting predicate was restricted by an internal call inside the dataflow library that ensures that the node being barrier'd is part of the path from a source to a sink, but this more robust 🙂